### PR TITLE
Functioning Layout for both Coins&Energy Bars

### PR DESCRIPTION
**Pull request by elShailee**
Made a functioning layout for both coins and energy bars, 

Including:
 - a button, a tab and a purchase window for each.
 - content placeholders.

Using: 
 - components for each one, general tab component, general close component.
 - a context which track and handle mounted components.

### CODE REVIEW

File: src/Components/Character/CharacterInfo.js [hunk 1]
```diff
@@ -4,9 +4,9 @@ import './CharacterInfo.css';
 export default function CharacterInfo() {
   return (
     <div id="characterIntoContainer">
```
Comment (EyalEliyahu) [inline src/Components/Character/CharacterInfo.js:6]: typo

File: src/Components/Coins/CoinsButton.js [hunk 2]
```diff
@@ -1,6 +1,32 @@
-import React from 'react';
+import React, { useContext } from 'react';
+import CoinsTab from './CoinsTab';
 import './CoinsButton.css';
+import { Button } from '@material-ui/core';
+import { MountedComponentsContext } from '../../Contexts/MountedComponentsContext';
```
Comment (EyalEliyahu) [inline src/Components/Coins/CoinsButton.js:5]: please change to `/Contexts/MountedComponentsContext`

File: src/Components/Coins/CoinsButton.js [hunk 3]
```diff
@@ -1,6 +1,32 @@
...
 
 export default function CoinsButton() {
-  return <div id="coinsButton">Coins</div>;
+  const { componentsList, ToggleOn, ToggleOff } = useContext(MountedComponentsContext);
+
+  let isCoinsTabOpen = componentsList.secondaries.CoinsTab;
```
Comment (EyalEliyahu) [inline src/Components/Coins/CoinsButton.js:10]: please move to const

File: src/Components/Coins/CoinsButton.js [hunk 4]
```diff
@@ -1,6 +1,32 @@
...
+      ToggleOn(['CoinsTab']);
+    }
+  };
+
+  return (
+    <div id="coinsButtonContainer">
+      <Button onClick={() => handleClick(isCoinsTabOpen)} id="coinsButton">
+        Coins
+      </Button>
+      {togglableContent}
```
Comment (EyalEliyahu) [inline src/Components/Coins/CoinsButton.js:29]: `{isConinsTabOpen && (<CoinsTab handleClose={() => ToggleOff(['CoinsTab', 'CoinsPurchaseWindow'])} />)}`
please use this intead of `toggbleContent`

File: src/Components/Coins/CoinsButton.js [hunk 5]
```diff
@@ -1,6 +1,32 @@
...
+    if (ifMounted) {
+      ToggleOff(['CoinsTab', 'CoinsPurchaseWindow']);
+    } else {
+      ToggleOn(['CoinsTab']);
+    }
+  };
+
+  return (
+    <div id="coinsButtonContainer">
+      <Button onClick={() => handleClick(isCoinsTabOpen)} id="coinsButton">
```
Comment (EyalEliyahu) [inline src/Components/Coins/CoinsButton.js:26]: `handleClick` can get `isCoinsTabOpen` by itself, it doesn't need to be passed as arg

File: src/Components/Coins/CoinsPurchaseWindow.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
+import React from 'react';
+import { Paper } from '@material-ui/core';
+import GeneralWindow from '../GeneralComponents/GeneralWindow/GeneralWindow';
```
Comment (EyalEliyahu) [inline src/Components/Coins/CoinsPurchaseWindow.js:3]: use global import instead of the `../`

File: src/Components/GeneralComponents/GeneralWindow/GeneralWindow.js [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    padding: '7px',
+    borderRadius: '5px',
+    zIndex: '1',
+  };
+
+  const centeredWindowStyle = {
+    position: 'fixed',
+    left: '50%',
+    top: '50%',
+    transform: 'translate(-50%,-50%)',
```
Comment (EyalEliyahu) [inline src/Components/GeneralComponents/GeneralWindow/GeneralWindow.js:17]: please add a comment for this line

File: src/Components/GeneralComponents/GeneralWindow/GeneralWindow.js [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+  const rightlyPositionedWindowStyle = {
+    position: 'fixed',
+    right: '7%',
+    top: '50%',
+    transform: 'translate(-50%,-50%)',
+  };
+
+  const composedWindowStyle = Object.assign(
+    basicWindowStyle,
+    props.windowPosition === 'right' ? rightlyPositionedWindowStyle : centeredWindowStyle
```
Comment (EyalEliyahu) [inline src/Components/GeneralComponents/GeneralWindow/GeneralWindow.js:29]: please use distracturing for `props.windowPosition`

Comment (EyalEliyahu) [reply src/Components/GeneralComponents/GeneralWindow/GeneralWindow.js:29]: please use const for the bool and const for the selected window style that is base on this bool.

File: src/Components/GeneralComponents/GeneralWindow/GeneralWindow.js [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+  };
+
+  const rightlyPositionedWindowStyle = {
+    position: 'fixed',
+    right: '7%',
+    top: '50%',
+    transform: 'translate(-50%,-50%)',
+  };
+
+  const composedWindowStyle = Object.assign(
```
Comment (EyalEliyahu) [inline src/Components/GeneralComponents/GeneralWindow/GeneralWindow.js:27]: please change `composedWindowStyle`  to `customizedWindowStyle`

File: src/Components/GeneralComponents/GeneralWindow/GeneralWindow.js [hunk 10]
```diff
@@ -0,0 +1,40 @@
+import React from 'react';
+import GeneralCloseButton from '../GeneralCloseButton/GeneralCloseButton';
+
+export default function GeneralWindow(props) {
```
Comment (EyalEliyahu) [inline src/Components/GeneralComponents/GeneralWindow/GeneralWindow.js:4]: lets use `func({children, handleClose})`

File: src/Components/GeneralComponents/GeneralWindow.js [hunk 11]
```diff
@@ -0,0 +1,47 @@
+import React from 'react';
+import GeneralCloseButton from 'Components/GeneralComponents/GeneralCloseButton';
+
+export default function GeneralWindow({ windowPosition, children, handleClose }) {
+  const isWindowRightSided = windowPosition === 'right' || windowPosition === 'Right';
```
Comment (EyalEliyahu) [inline src/Components/GeneralComponents/GeneralWindow.js:5]: inside `consts/generalStyles.js`
const windowPositionsOptions = {
right,
left,
top
bottom
}
on `generalWindow.js` :
`const isWindowRightSided = windowPosition === windowPositionsOptions.right`

on `asd.js` (that uses generalWindow):
<GeneralWidnow windowPositon={windowPositionsOptions.right}>

File: src/Components/Seen/SeenButton.css [hunk 12]
```diff
@@ -1,12 +1,11 @@
...
   background-color: cornflowerblue;
   width: 60px;
   height: 60px;
   border-radius: 50%;
-  top: 0;
-  right: 0px;
-  border: 1px solid black;
+  margin-bottom: 20px;
+  margin-left: 60px;
+  border: 0px solid black;
```
Comment (EyalEliyahu) [inline src/Components/Seen/SeenButton.css:8]: check if it's needed, black is useless

File: src/Contexts/MountedComponentsContext.js [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+        CoinsPurchaseWindow: false,
+      },
+      secondaries: {
+        CoinsTab: false,
+        EnergyTab: false,
+      },
+    },
+  };
+
+  ComponentsMounterAndUnmounter = (componentsToMountArray, isToMount) => {
```
Comment (EyalEliyahu) [inline src/Contexts/MountedComponentsContext.js:21]: `ComponentsMounterAndUnmounter` => `mountOrUnmountComponents`

Comment (EyalEliyahu) [reply src/Contexts/MountedComponentsContext.js:21]: `isToMount` => `shouldMount`

File: src/Contexts/MountedComponentsContext.js [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+    const mountedComponentsDict = this.state.mountedComponentsDict;
+    const { ...priorityGroups } = mountedComponentsDict;
+    componentsToMountArray.forEach(componentToMountOrUnmount => {
+      for (const group in priorityGroups) {
```
Comment (EyalEliyahu) [inline src/Contexts/MountedComponentsContext.js:25]: `group` => `priorityGroup`

File: src/Contexts/MountedComponentsContext.js [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+        if (componentToMountOrUnmount in groupInDict) {
+          const { mountedComponentsDict: newMountedComponentsDict } = this.state;
+          newMountedComponentsDict[group][componentToMountOrUnmount] = isToMount;
+          this.setState({ mountedComponentsDict: newMountedComponentsDict });
+        }
+      }
+    });
+  };
+
+  MountComponents = componentsToUnmountArray => {
```
Comment (EyalEliyahu) [inline src/Contexts/MountedComponentsContext.js:36]: camel case

File: src/Contexts/MountedComponentsContext.js [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+    this.ComponentsMounterAndUnmounter(componentsToUnmountArray, true);
+  };
+  UnmountComponents = componentsToUnmountArray => {
+    this.ComponentsMounterAndUnmounter(componentsToUnmountArray, false);
+  };
+
+  render() {
+    return (
```
Comment (EyalEliyahu) [inline src/Contexts/MountedComponentsContext.js:44]: const { state, MountComponents, ... } = this;

File: src/Texts/Texts.js [hunk 17]
```diff
@@ -0,0 +1,10 @@
+const texts = {
+  logs: {
```
Comment (EyalEliyahu) [inline src/Texts/Texts.js:2]: please change `logs.errors` to errorMesseges

Comment (elShailee) [reply src/Texts/Texts.js:2]: moved only this error text to AppUtils, in order to be able to individually export texts's text categories, due to IE Breaking if texts had exports which wasn't default.

File: src/App/AppUtils.js [hunk 18]
```diff
@@ -0,0 +1,14 @@
...
+  // Edge 20-78 (not Chromium based)
+  const isOldEdge = !isIE && !!window.StyleMedia;
+
+  const isInvalidBrowser = isIE || isOldEdge;
+
+  return !isInvalidBrowser;
+}
+
+export const invalidBrowserMessage =
+  'This Application is not supported for this browser, please try to open it using an up-to-date version of Google Chrome, Mozilla Firefox, or Microsoft Edge.';
```
Comment (elShailee) [inline src/App/AppUtils.js:14]: move to texts

File: src/App/App.js [hunk 19]
```diff
@@ -0,0 +1,10 @@
+import React from 'react';
+import GameplayScreen from 'Screens/GameplayScreen/GameplayScreen';
+import { isBrowserValid, invalidBrowserMessage } from './AppUtils';
+
+export default function App() {
+  const isBrowserValidBool = isBrowserValid();
+  if (isBrowserValidBool) {
```
Comment (elShailee) [inline src/App/App.js:7]: all this logic should all be a function, App.js shouldn't know the mechanics of the valid-browser check and what message to display.

File: src/Texts/Texts.js [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+  },
+  energy: {
+    energySymbol: '♠',
+    fullIn: 'Full Energy in:',
+    rate: 'Rate: ',
+    energyPerHour: '♠/h',
+  },
+};
+export default texts;
+export const { componentsTitles, buttonsText, energy, coins } = texts;
```
Comment (elShailee) [inline src/Texts/Texts.js:28]: should put more thought into the hirarchi of this element and how to export it.
maybe it should be built around fields of context, such as ui vs content, etc.
maybe by compontns in code?
the goal is that it'll be more intuitive to be imported and used in the components.

File: src/Texts/Texts.js [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+  },
+  buttonsText: {
+    getMore: 'Get More',
+  },
+  coins: {
+    amountPt1: 'You have ',
+    amountPt2: 'Coins.',
+  },
+  energy: {
+    energySymbol: '♠',
```
Comment (elShailee) [inline src/Texts/Texts.js:21]: symbol shouln't be in text, add a placeholder icon

File: src/Components/Energy/EnergyTab.js [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+    <Card id="energyTab">
+      <Typography variant={'subtitle2'} className={classes.energyInfo} id="timeUntillFullText">
+        {energy.fullIn}
+        <br />
+        XX:XX:XX
+      </Typography>
+
+      <Typography variant={'subtitle2'} className={classes.energyInfo} id="energyAmountText">
+        XX
+        {energy.energySymbol}
```
Comment (elShailee) [inline src/Components/Energy/EnergyTab.js:52]: symbol shouln't be in text, add a placeholder icon

File: src/Components/Energy/EnergyTab.js [hunk 23]
```diff
@@ -0,0 +1,74 @@
...
+      </Typography>
+
+      <Typography variant={'subtitle2'} className={classes.energyInfo} id="rechargeRateText">
+        {energy.rate}
+        XX
+        {energy.energyPerHour}
```
Comment (elShailee) [inline src/Components/Energy/EnergyTab.js:58]: symbol shouln't be in text, add a placeholder icon

File: src/Components/Actions/ActionsButton.js [hunk 24]
```diff
@@ -1,6 +1,11 @@
 import React from 'react';
 import './ActionsButton.css';
+import { componentsTitles } from 'Texts/texts';
 
 export default function ActionButton() {
-  return <div id="actionsButton">Actions</div>;
+  return (
+    <div id="upperRightButtonsContainer">
```
Comment (elShailee) [inline src/Components/Actions/ActionsButton.js:7]: is this div necessary or in the right place?
also - change it's name to be more descriptive, and move it's css to the place it exists.

File: src/Components/Coins/CoinsButton.js [hunk 25]
```diff
@@ -1,6 +1,29 @@
...
 
 export default function CoinsButton() {
-  return <div id="coinsButton">Coins</div>;
+  const { mountedComponentsDict, mountComponents, unmountComponents } = useContext(MountedComponentsContext);
+
+  const isCoinsTabOpen = mountedComponentsDict.secondaries.CoinsTab;
+
+  const onCoinsButtonClick = () => {
+    if (isCoinsTabOpen) {
+      unmountComponents(['CoinsTab', 'CoinsPurchaseWindow']);
```
Comment (elShailee) [inline src/Components/Coins/CoinsButton.js:15]: see with eyal if this is the best practice (to unmount CoinsPurchaseWindow in any case)

File: src/Components/Character/CharacterInfo.css [hunk 26]
```diff
@@ -2,10 +2,8 @@
   width: 400px;
   height: 100px;
   background-color: darkblue;
-  position: absolute;
   text-align: center;
   clip-path: polygon(0 20px, 20px 0, 400px 0, 360px 60px, 180px 60px, 160px 100px, 40px 100px);
-  z-index: 4;
 }
 
 #xpCapacity {
```
Comment (elShailee) [inline src/Components/Character/CharacterInfo.css:9]: these two are almost identical, can't they be combine in some ways?

File: src/Components/Character/CharacterInfo.css [hunk 27]
```diff
@@ -27,5 +24,10 @@
   top: 60px;
   left: 170px;
   clip-path: polygon(0 0, 160px 0, 153px 10px, 0 10px);
-  z-index: 2;
+}
+
+#characterInfoContainer {
+  position: absolute;
+  left: calc(50% - 200px);
```
Comment (elShailee) [inline src/Components/Character/CharacterInfo.css:31]: is there a way for this 200px to be defined as half of the 400px above?

File: src/Components/Character/CharacterInfo.css [hunk 28]
```diff
@@ -27,5 +24,10 @@
...
+  top: calc(8px);
```
Comment (elShailee) [inline src/Components/Character/CharacterInfo.css:32]: no need for calc.

File: src/Contexts/MountedComponentsContext.js [hunk 29]
```diff
@@ -0,0 +1,52 @@
...
+        CoinsPurchaseWindow: false,
+      },
+      secondaries: {
+        CoinsTab: false,
+        EnergyTab: false,
+      },
+    },
+  };
+
+  mountOrunmountComponents = (componentsToMountArray, shouldMount) => {
```
Comment (elShailee) [inline src/Contexts/MountedComponentsContext.js:21]: - update accordingly to removal of categories
 - please add a check for componentsToMountArray to actually be an array

File: src/Contexts/MountedComponentsContext.js [hunk 30]
```diff
@@ -0,0 +1,52 @@
+import React, { Component, createContext } from 'react';
+
+export const MountedComponentsContext = createContext();
+
+class MountedComponentsContextProvider extends Component {
+  state = {
+    mountedComponentsDict: {
```
Comment (elShailee) [inline src/Contexts/MountedComponentsContext.js:7]: there is no use for the primaries and secondaries categories

File: src/Components/Energy/EnergyButton.js [hunk 31]
```diff
@@ -1,6 +1,30 @@
...
+import { componentsTitles } from 'Texts/texts';
 
 export default function EnergyButton() {
-  return <div id="energyButton">Energy</div>;
+  const { mountedComponentsDict, mountComponents, unmountComponents } = useContext(MountedComponentsContext);
+  const isEnergyTabOpen = mountedComponentsDict.secondaries.EnergyTab;
+
+  const onEnergyButtonClick = ifMounted => {
+    if (isEnergyTabOpen) {
+      unmountComponents(['EnergyTab', 'EnergyPurchaseWindow']);
```
Comment (elShailee) [inline src/Components/Energy/EnergyButton.js:14]: same as in CoinsButton.js

File: src/Components/GeneralComponents/GeneralTab.js [hunk 32]
```diff
@@ -0,0 +1,5 @@
+import React from 'react';
+
+export default function GeneralTab({ children }) {
+  return <div style={{ width: 'fit-content' }}>{children}</div>;
```
Comment (elShailee) [inline src/Components/GeneralComponents/GeneralTab.js:4]: please move style to css file

File: src/Components/GeneralComponents/GeneralWindow.js (hunk 11 above)
Comment (elShailee) [reply src/Components/GeneralComponents/GeneralWindow.js:5]: maybe make as a function on this component? as this is only relevant to this component.

File: src/Components/GeneralComponents/GeneralWindow.js [hunk 33]
```diff
@@ -0,0 +1,47 @@
...
+
+  const customizedWindowStyle = Object.assign(
+    basicWindowStyle,
+    isWindowRightSided ? rightlyPositionedWindowStyle : centeredWindowStyle
+  );
+
+  return (
+    <div style={customizedWindowStyle}>
+      <GeneralCloseButton unmountWindow={unmountGeneralWindow} />
+      <div id="contentContainer" style={{ marginTop: '7px' }}>
```
Comment (elShailee) [inline src/Components/GeneralComponents/GeneralWindow.js:42]: move style to css file

File: src/Components/Map/Map.css [hunk 34]
```diff
@@ -8,6 +8,5 @@
   left: 20px;
   width: 180px;
   height: 340px;
-  z-index: 2;
-  background-color: rgb(255, 255, 255, 0);
+  z-index: 1;
```
Comment (elShailee) [inline src/Components/Map/Map.css:11]: why is this necessary?

File: src/Components/Map/MapCompass.css [hunk 35]
```diff
@@ -6,7 +6,6 @@
   height: 160px;
   width: 160px;
   background: #333;
-  border-style: 1px solid black;
   line-height: 160px;
```
Comment (elShailee) [inline src/Components/Map/MapCompass.css:9]: can this be a variable that relates to the height?

File: src/Components/Seen/SeenButton.css [hunk 36]
```diff
@@ -1,12 +1,10 @@
...
   height: 60px;
   border-radius: 50%;
-  top: 0;
-  right: 0px;
-  border: 1px solid black;
+  margin-bottom: 20px;
+  margin-left: 60px;
 
   text-align: center;
   line-height: 60px;
```
Comment (elShailee) [inline src/Components/Seen/SeenButton.css:10]: can this be a variable that relates to the height?

File: src/index.js [hunk 37]
```diff
@@ -1,7 +1,9 @@
+import 'react-app-polyfill/ie11';
+import 'react-app-polyfill/stable';
```
Comment (elShailee) [inline src/index.js:2]: what are those?
are they necessary?

File: src/Components/Coins/CoinsButton.js (hunk 25 above)
Comment (elShailee) [reply src/Components/Coins/CoinsButton.js:15]: nah it's cool

File: src/Components/Character/CharacterInfo.css (hunk 28 above)
Comment (elShailee) [reply src/Components/Character/CharacterInfo.css:32]:
```suggestion
  top: 8px;
```

File: src/Components/Map/Map.css (hunk 34 above)
Comment (elShailee) [reply src/Components/Map/Map.css:11]: wasn't
